nvtabular/ops/categorify.py

```diff
@@ -753,13 +775,17 @@ def _encode(name, storage_name, path, gdf, cat_cache, na_sentinel=-1, freq_thres
         ).sort_values("order")["labels"]
         labels.fillna(na_sentinel, inplace=True)
         labels = labels.values
-    else:
-        # Use `searchsorted` if we are using a "full" encoding
+    if search_sorted:
+        # Use `searchsorted` if doing full encoding (optional).
         if list_col:
             labels = value[selection_r].searchsorted(
                 gdf[selection_l[0]].list.leaves, side="left", na_position="first"
             )
         else:
+            for c in selection_l:
+                typ = gdf[c].dtype
+                if typ != value[c].dtype:
+                    value[c] = value[c].astype(typ)
```

Comment:
Is this for the problem of int columns with nulls getting converted to floats? If so - I'd rather solve in a way that doesn't require data being cast multiple times (like from int->float->int).

Rick started a thread on the cudf about this (https://github.com/rapidsai/cudf/issues/6536) - I think we probably want to spill to arrow rather than pandas (which would be out of scope of this PR

Reply:
Right - I told @rnyak I would take care of it separately, but may make sense to include the small change here.

The idea is to replace all instances (in `categorify.py`) of

```python
df = _concat(dfs, ignore_index)
if on_host:
    df.reset_index(drop=True, inplace=True)
    df = cudf.from_pandas(df)
```

with

```python
if on_host:
    df = pa.concat_tables(dfs)
    df = cudf.DataFrame.from_arrow(df)
else:
    df = _concat(dfs, ignore_index)
```
and all instances of

```python
.to_pandas()
```

to

```python
.to_arrow(preserve_index=False)
```

(where `pa` is `pyarrow`). 

Does this sound reasonable?

Reply:
@benfred  what are we deciding here? shall we fix `nulls getting converted to floats` in another PR? should I leave this part of code as it is?

Reply:
Lets do in a separate PR - this is unrelated to the searchsorted issue afaict. 

This code shouldn't be part of the new PR though (and should just be removed from this PR). The new PR should use Rick's suggestions